src/main/java/com/microsoft/sqlserver/jdbc/SQLServerMSAL4JUtils.java

```diff
@@ -119,6 +127,76 @@ static SqlFedAuthToken getSqlFedAuthTokenIntegrated(SqlFedAuthInfo fedAuthInfo,
+            IAuthenticationResult authenticationResult = null;
+
+            // try to acquire token silently if user account found in cache
+            try {
+                Set<IAccount> accountsInCache = pca.getAccounts().join();
+                if (null != accountsInCache && !accountsInCache.isEmpty() && null != user && !user.isEmpty()) {
+                    IAccount account = getAccountByUsername(accountsInCache, user);
+                    if (null != account) {
+                        if (logger.isLoggable(Level.FINE)) {
+                            logger.fine(logger.toString() + "Silent authentication for user:" + user);
+                        }
+                        SilentParameters silentParameters = SilentParameters
+                                .builder(Collections.singleton(fedAuthInfo.spn + "/.default"), account).build();
+
+                        future = pca.acquireTokenSilently(silentParameters);
+                    }
+                }
+            } catch (MsalInteractionRequiredException e) {
+                // not an error, need to get token interactively
+            }
+
+            if (null != future) {
+                authenticationResult = future.get();
+            } else {
+                // acquire token interactively with system browser
+                if (logger.isLoggable(Level.FINE)) {
+                    logger.fine(logger.toString() + "Interactive authentication");
+                }
+                InteractiveRequestParameters parameters = InteractiveRequestParameters.builder(new URI(REDIRECTURI))
+                        .systemBrowserOptions(SystemBrowserOptions.builder()
+                                .htmlMessageSuccess(SQLServerResource.getResource("R_MSALAuthComplete")).build())
+                        .loginHint(user).scopes(Collections.singleton(fedAuthInfo.spn + "/.default")).build();
+
+                future = pca.acquireToken(parameters);
+                authenticationResult = future.get();
+            }
+
+            return new SqlFedAuthToken(authenticationResult.accessToken(), authenticationResult.expiresOnDate());
+        } catch (MalformedURLException | InterruptedException | URISyntaxException e) {
+            throw new SQLServerException(e.getMessage(), e);
+        } catch (ExecutionException e) {
+            throw getCorrectedException(e, user, authenticationString);
+        } finally {
+            executorService.shutdown();
+        }
+    }
+
+    // Helper function to return account containing user name from set of accounts, or null if no match
+    private static IAccount getAccountByUsername(Set<IAccount> accounts, String username) {
+        if (!accounts.isEmpty()) {
+            for (IAccount account : accounts) {
+                if (account.username().equals(username)) {
+                    return account;
+                }
+            }
+        }
+        return null;
+    }
```

Comment:
Are null checks not required for accounts, entries, or usernames? The following checks if `accounts` is null and not empty, before attempting to find the first element that matches the username. It also checks that the `IAccount` it just iterated over isn't null, the `username` within the `IAccount` isn't null, and then does the comparison. If it finds an element that matches the criteria, it'll assign that value to the return variable, if not, it'll assign `null` to the return variable.

```java
// Helper function to return account containing user name from set of accounts, or null if no match
private static IAccount getAccountByUsername(Set<IAccount> accounts, String username) {
    IAccount result = null;
    if (null != accounts && !accounts.isEmpty())
        result = accounts.parallelStream()
                .filter(a -> (null != a && null != a.username() && a.username().equals(username))).findFirst()
                .orElse(null);
    return result;
}
```

Reply:
accounts is the set of accounts in cache, which could be and would be null on 1st connection.